File: sdk/python/qlix/tool_profiles.py
"""OpenClaw-style tool profiles — filter effective scopes before ToolRouter builds the catalog."""

from __future__ import annotations

_MINIMAL = frozenset(
    {
        "web.read",
        "web.research",
        "brain.query",
        "brain.knowledge_read",
    }
)
_CODING = _MINIMAL | frozenset(
    {
        "web.click",
        "system.file_read",
        "system.file_write",
    }
)


def filter_scopes_by_tool_profile(scopes: list[str], profile: str | None) -> list[str]:
    """Narrow granted scopes by profile.

    ``lean`` is not a scope filter — it keeps every scope but drops rarely-used
    management tools from the schema (see tool_budget.ADMIN_TOOLS). It is applied
    later, on the assembled tool list, so it works uniformly across connectors.
    """
    p = (profile or "full").strip().lower()
    if p in ("full", "lean") or not p:
        return list(scopes)
    allow = _MINIMAL if p == "minimal" else _CODING
    return [s for s in scopes if s in allow or str(s).startswith("mcp.")]

File: backend/src/agents/toolProfiles.ts
import type { PermissionScope } from './agents.types.js';

/**
 * `lean` keeps every granted scope but drops rarely-used management tools (create a
 * channel, bulk CRM ops) from the schema sent each round. Unlike `minimal`/`coding`
 * it is not a scope filter, so it never changes what the agent is permitted to do —
 * only what it is shown by default. Applied in the SDK's tool_budget module.
 */
export type ToolProfile = 'minimal' | 'coding' | 'lean' | 'full';

const MINIMAL_SCOPES: ReadonlySet<string> = new Set([
  'web.read',
  'web.research',
  'brain.query',
  'brain.knowledge_read',
]);

const CODING_SCOPES: ReadonlySet<string> = new Set([
  ...MINIMAL_SCOPES,
  'web.click',
  'system.file_read',
  'system.file_write',
]);

/**
 * Filter granted scopes by OpenClaw-style tool profile.
 * `full` returns the input unchanged.
 */
export function filterScopesByToolProfile(
  scopes: PermissionScope[],
  profile: ToolProfile | string | null | undefined,
): PermissionScope[] {
  const p = (profile ?? 'full').toLowerCase() as ToolProfile;
  // `lean` narrows the tool schema, not the permission set — see the type doc above.
  if (p === 'full' || p === 'lean') return scopes;
  const allow = p === 'minimal' ? MINIMAL_SCOPES : CODING_SCOPES;
  return scopes.filter((s) => allow.has(s) || s.startsWith('mcp.'));
}

export function isValidToolProfile(value: unknown): value is ToolProfile {
  return value === 'minimal' || value === 'coding' || value === 'lean' || value === 'full';
}
